### ISSUE TITLE
Per-transaction triggers

### ISSUE THREAD

**Issue by mversic** [hyperledger-iroha/iroha#4937]
Current execution model is this:
1) isi produce events (including ExecuteTrigger event)
2) triggers are matched against events and put into the queue for later execution
3) after block commit they are executed in the matched order

Consider this case:
1. register trigger that is activated on Account::MetadataInsert
2. wait for block to be committed

3. insert "key" into "alice" metadata
4. mint "rose" for "alice"
5. insert "key" into "alice" metadata
6. wait for block to be committed

In this case trigger is executed twice at the end of the block instead of when actually triggered. This means that by the time trigger reacts to an event, the event is stale. 

IMO this execution model is weird and not really reliable.

**Comment by mversic** [comment]
related to #4602

**Comment by Mingela** [comment]
yeah it seems trigger executions should rather be injected immediately after a related event is emitted

**Comment by s8sato** [comment]
Core challenges:

- mitigate the performance overhead of matching events per instruction
- address potential execution loops of data triggers

**Comment by s8sato** [comment]
Now I'm going with a policy of matching events on transactions, not instructions. Since after #4968 trigger execution will be a transaction, it cannot be inserted between instructions. So the ordering problem will be mitigated (per transaction) rather than completely fixed (per instruction)

**Comment by s8sato** [comment]
To prevent infinite chain of trigger executions, the following two approaches can be considered:

1. On execution, track executed triggers and dismiss the subsequent matches
2. On registration, reject the trigger registration if event loop detected

1 is a relative easy approach to implement, but the implicit cancellation may be surprising to users even though the execution result is deterministic.
2 guarantees no loop at registration, which requires (possible) events being able to be derived from instructions regardless of state in order to detect loops in advance.

I've been working with approach 2, but now I'm wondering if approach 1 would be accepted

**Comment by s8sato** [comment]
A hard problem is that the current Wasm executables are entirely unpredictable, meaning there is no guarantee that data triggers will not enter an infinite execution loop. Tracking issue:

- #5356

**Comment by s8sato** [comment]
Regardless of whether we take a registration-time or execution-time approach for loop detection, the fundamental issue with data triggers is their lack of validation when subscribing to events.
One of the easiest ways to break all data triggers is to register a trigger that subscribes to any event, which alone creates a cyclic graph.

My suggestion is to introduce read access control for event subscriptions, similar to queries:

- #5338

**Comment by mversic** [comment]
This is a known issue and has been discussed in #4429 to some extent

**Comment by s8sato** [comment]
# Trigger Safety Mechanism: Implementation Plan

To prevent infinite execution loops caused by recursively activated triggers, I propose a phased rollout of execution control mechanisms as follows:

## Phase 1: Step Limit Enforcement

- Impose a maximum execution depth for chained triggers within a single transaction.
- Any trigger executions beyond this limit are aborted.
- Ensures basic safety with minimal implementation complexity.

## Phase 2: Trigger Scope Restriction via Subscribe Permissions

- Introduce "subscribe permissions" to control which state changes a trigger can listen to.
- Prevents unintended activations and limits cross-account entanglement.

## Phase 3: Execution Points Quota (Solana-style Model)

- Introduce a virtual quota system (execution points) per transaction.
- Each trigger consumes a predefined amount of points; execution halts once the quota is exhausted.
- The sender of the original transaction bears the cost.
- For human accounts, execution points can be converted into token-based fees to provide economic control.

## Future Consideration: Static Loop Detection via Path Exhaustion

- In PR #5355, we considered performing pessimistic static analysis at trigger registration to detect potential recursive loops across all execution paths.
- While this would ensure loop-free behavior, it may also reject valid triggers due to over-approximation.
- Setting this option aside for now allows decoupling the issue from #5362 and #5357 in the dependency graph (#5356).

**Comment by s8sato** [comment]
Phase 3:

> - Each trigger consumes a predefined amount of points; execution halts once the quota is exhausted.

Should we ensure that the entire sequence -- starting from the initial cause (an external transaction or a time trigger) to the subsequent internal "transactions" -- is treated as a single transactional unit?
